pkg/ipmasq/ipmasq.go

```diff
@@ -99,39 +97,35 @@ type IPMasqAgent struct {
 	handlerFinished        chan struct{}
 }
 
-func NewIPMasqAgent(logger *slog.Logger, reg *metrics.Registry, configPath string) (*IPMasqAgent, error) {
-	return newIPMasqAgent(logger, configPath, &ipmasq.IPMasqBPFMap{MetricsRegistry: reg})
-}
-
-func newIPMasqAgent(logger *slog.Logger, configPath string, ipMasqMap IPMasqMap) (*IPMasqAgent, error) {
-	watcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		return nil, fmt.Errorf("Failed to create fsnotify watcher: %w", err)
-	}
-
-	configDir := filepath.Dir(configPath)
-	// The directory of the config should exist at this time, otherwise
-	// the watcher will fail to add
-	if err := watcher.Add(configDir); err != nil {
-		watcher.Close()
-		return nil, fmt.Errorf("Failed to add %q dir to fsnotify watcher: %w", configDir, err)
-	}
-
+func NewIPMasqAgent(logger *slog.Logger, configPath string, ipMasqMap IPMasqMap) *IPMasqAgent {
 	a := &IPMasqAgent{
 		logger:                 logger,
 		configPath:             configPath,
 		nonMasqCIDRsFromConfig: map[string]netip.Prefix{},
 		nonMasqCIDRsInMap:      map[string]netip.Prefix{},
 		ipMasqMap:              ipMasqMap,
-		watcher:                watcher,
 	}
 
-	return a, nil
+	return a
 }
 
 // Start starts the ip-masq-agent goroutine which tracks the config file and
 // updates the BPF map accordingly.
-func (a *IPMasqAgent) Start() {
+func (a *IPMasqAgent) Start() error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create fsnotify watcher: %w", err)
+	}
+	a.watcher = watcher
```

Comment:
I guess you can move this initialization into the constructor? It doesn't involve any IO blocking right?

Reply:
The watcher does need to be closed, so perhaps it makes sense to have it as part of `Start`, since we don't expect the constructor to have side effects that need to be closed.